Add blog post about the current C# platform support in 4.2

raulsntos opened the pull request:
A quick progress update on platform support for C# projects. The intention of the article is to answer these frequent questions regarding C# support:

- What platforms does Godot currently support?
- Why did mobile/web platform support disappear in the move from 3.x to 4.0?
- Why is mobile support considered experimental?
- Why is the web platform still unsupported?
- What are the next steps? Or what is missing to get full mobile/web support?

I didn't want the article to get too technical, but if you think these questions aren't sufficiently answered please let me know.

raulsntos committed:
Add blog post about the current C# platform support in 4.2

raulsntos commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+excerpt: "Progress update about the platforms supported by C# projects as of Godot 4.2."
+categories: ["progress-report"]
+author: Raul Santos
+image: /storage/blog/covers/platform-state-in-csharp-for-godot-4-2.jpg
```
TODO: I don't have a cover image.

raulsntos commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+categories: ["progress-report"]
+author: Raul Santos
+image: /storage/blog/covers/platform-state-in-csharp-for-godot-4-2.jpg
+date: 2023-11-25 17:00:00
```
TODO: Date is TBD.

raulsntos commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+As a summary, the current platform support as of 4.2 is described in the table below:
+
+| Platform | Runtimes supported | Minimum required .NET version |
+| -: | - | - |
+| **Windows** | CoreCLR, Mono, NativeAOT | 6.0 (CoreCLR, Mono), 7.0 (NativeAOT) |
+| **macOS** | CoreCLR, Mono, NativeAOT | 6.0 (CoreCLR, Mono), 7.0 (NativeAOT) |
+| **Linux** | CoreCLR, Mono, NativeAOT | 6.0 (CoreCLR, Mono), 7.0 (NativeAOT) |
+| **Android** | Mono | 7.0 |
+| **iOS** | NativeAOT | 8.0 |
+| **Web** | - | - |
```
The table doesn't look great :disappointed: 

![image](https://github.com/godotengine/godot-website/assets/3903059/4bc05da8-dd42-4b1e-9f48-aaa225a49734)

Calinou replied:
We should add blog table styles in a dedicated PR.

YuriSizov commented:
@raulsntos Is there anything you want to add here aside from the cover and the date and the table styling? You mentioned web exports being up in the air, so I guess that needs to be adjusted to follow the current status?

raulsntos commented:
I looked into implementing web exports but it didn't pan out, so the current status remains. Web exports are currently unimplemented and likely depend on upstream improvements.

I mainly left this as a draft in case I was able to get something working but since that didn't happen I'll mark it ready for review.

MurderVeggie commented:
I just read through the article and really like it, easy to understand language despite tech-heavy topic, well done!

I do have two questions however:
- the Web part lost me half way through; not supported yet but somehow it is? maybe some additional explanation on what you can/cannot do with GDextension as temporary solution would be helpful; a prognosis for when official support is inbound or describing the current blockers could be beneficial too
- hate to say it, but the first question we are going to get is "What about console platforms?" so maybe that warrants at least one sentence

raulsntos commented:
Thank you so much for the review.

> the Web part lost me half way through; not supported yet but somehow it is? maybe some additional explanation on what you can/cannot do with GDextension as temporary solution would be helpful; a prognosis for when official support is inbound or describing the current blockers could be beneficial too

I think I was trying too hard to be optimistic about the progress since I know a lot people ask about Web support. I only mentioned GDExtension because it's somewhat related but it doesn't really affect C# support directly so maybe it's better to avoid the mention.

The take-away should be that the Web platform is still unsupported and that, despite some progress being made, it may still take a while (which may require waiting for upstream).

In the last paragraph I also try to explain that .NET can build WASM but not in a way that can be used by Godot. I added this explanation because some people may be confused as to why the Web platform is still unsupported since they may be aware that .NET can build WASM (e.g.: [Blazor](https://dotnet.microsoft.com/en-us/apps/aspnet/web-apps/blazor)). But maybe this paragraph only made things more confusing.

So to try and answer your questions:

- You can't really use GDExtension as a workaround, C# is still not using GDExtension in any way. It's just that they way they are both implemented is somewhat similar, so it's fair to assume that improvements to one could be applied to the other.
- I'd rather avoid giving any specific deadlines, but I do mention at the end that .NET 9 may include what we need and that's November 2024 at the earliest.
- Current blockers would be, as mentioned in the last paragraph, that .NET doesn't allow building WASM as side-modules that can be loaded by Godot (there's also a link to the issue in the .NET repository so people can keep track of it).

> What about console platforms?

That's not an area I'm familiar with and not something that I've been working on. I believe this would be the responsability of W4 Games, I can ask around and see if anybody could add something.

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+First a bit of history. Godot 3 supports exporting C# projects to all the platforms supported by the engine. The C# implementation uses the [Mono embedding APIs](https://www.mono-project.com/docs/advanced/embedding/) and the [Mono runtime](https://www.mono-project.com/docs/advanced/runtime/). **Mono** is an open source cross-platform implementation of the Windows-only .NET Framework.
+
+With the [4.0 release](https://godotengine.org/article/godot-4-0-sets-sail/), the C# integration moved away from the Mono embedding APIs and replaced it with the .NET Core hosting APIs (using the `hostfxr` library). This allowed us to modernize the codebase and prepare for [.NET 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/), the first release in the .NET unification journey. Unfort ...
```
```suggestion
With the [4.0 release](https://godotengine.org/article/godot-4-0-sets-sail/), the C# integration moved away from the Mono embedding APIs and replaced it with the .NET Core hosting APIs (using the `hostfxr` library). This allowed us to modernize the codebase and prepare for [.NET 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/), the first release in the .NET unification journey. Unfortunately, in this move C# projects lost the ability to export to platforms other than Desktop (Windows, macOS and Linux).
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+The term **Mono** can be used to refer to a collection of technologies. With .NET unification, the Mono framework is deprecated, but the runtime is still supported. Unified .NET uses both the CoreCLR and the Mono runtimes, but Mono is the only runtime that supports mobile and web platforms.
+
+However, in .NET 7.0 a new runtime became available, [**NativeAOT**](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot), which allows publishing native binaries built for a specific platform, instead of the JIT-based portable binaries. In .NET 7.0, NativeAOT only supports Windows and Linux, but in .NET 8.0 it adds support for MacOS and experimental support for Android and iOS. Th ...
```
```suggestion
However, in .NET 7.0 a new runtime became available, [**NativeAOT**](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot), which allows publishing native binaries built for a specific platform, instead of the JIT-based portable binaries. In .NET 7.0, NativeAOT only supports Windows and Linux, but in .NET 8.0 it adds support for macOS and experimental support for Android and iOS. This means in .NET 8.0 NativeAOT can be used as an alternative to Mono for mobile platforms.
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+## Platform support in 4.2
+
+Thanks to the amazing work by [RedworkDE](https://github.com/RedworkDE) and [Andreia Gaita](https://github.com/shana), C# projects in Godot [4.2](/article/godot-4-2-arrives-in-style) have _experimental_ support for exporting to **Android** and **iOS**.
```
```suggestion
Thanks to amazing work by [RedworkDE](https://github.com/RedworkDE) and [Andreia Gaita](https://github.com/shana), C# projects in Godot [4.2](/article/godot-4-2-arrives-in-style) have _experimental_ support for exporting to **Android** and **iOS**.
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+Thanks to the amazing work by [RedworkDE](https://github.com/RedworkDE) and [Andreia Gaita](https://github.com/shana), C# projects in Godot [4.2](/article/godot-4-2-arrives-in-style) have _experimental_ support for exporting to **Android** and **iOS**.
+
+This initial implementation has some limitations which is why they are still marked experimental, we want to make sure the platform support is production-ready when we remove the experimental label. The platform limitations are documented in the [C# documentation](https://docs.godotengine.org/en/4.2/tutorials/scripting/c_sharp/index.html#c-platform-support), but we'll go over each platform here.
```
```suggestion
These initial implementations have some limitations which is why they are still marked experimental. We want to make sure the platform support is production-ready before we remove the experimental label. The platform limitations are documented in the [C# documentation](https://docs.godotengine.org/en/4.2/tutorials/scripting/c_sharp/index.html#c-platform-support), but we'll go over each platform here.
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+Exporting to Android requires .NET **7.0** or higher, uses the [**linux-bionic**](https://github.com/dotnet/runtime/pull/66147) Mono runtime, and only supports the `arm64` and `x64` architectures.
+
+The **linux-bionic** runtime is a linux runtime using the Android C library (so it's basically Android but without the JNI). This means that the Android bindings are not available, so some APIs (such as [SSL](https://github.com/godotengine/godot/issues/84559)) will crash the game, use the Godot APIs when possible to avoid these issues.
```
```suggestion
The **linux-bionic** runtime is a Linux runtime using the Android C library, so it's basically Android but without the JNI. This means that the Android bindings are not available, so some APIs (such as [SSL](https://github.com/godotengine/godot/issues/84559)) will crash the game. Use the Godot APIs when possible to avoid these issues.
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+The **linux-bionic** runtime is a linux runtime using the Android C library (so it's basically Android but without the JNI). This means that the Android bindings are not available, so some APIs (such as [SSL](https://github.com/godotengine/godot/issues/84559)) will crash the game, use the Godot APIs when possible to avoid these issues.
+
+Using NativeAOT should also be supported in theory, but it requires using .NET **8.0** and some manual work, you can read more about using NativeAOT with Android bionic in the [Microsoft documentation](https://github.com/dotnet/runtime/blob/v8.0.0/src/coreclr/nativeaot/docs/android-bionic.md).
```
```suggestion
Using NativeAOT should also be supported in theory, but it requires using .NET **8.0** and some manual work. You can read more about using NativeAOT with Android bionic in the [Microsoft documentation](https://github.com/dotnet/runtime/blob/v8.0.0/src/coreclr/nativeaot/docs/android-bionic.md).
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+NativeAOT uses trimming and Godot still uses reflection in parts of the codebase, so we are rooting the Godot bindings and the game project's assemblies but some reflection usage may still break at runtime.
+
+Exporting to iOS is only supported from a MacOS device and requires using Xcode to build the final binaries.
```
```suggestion
Exporting to iOS is only supported from a macOS device and requires using Xcode to build the final binaries.
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+The web platform is still **not supported** in 4.2, but progress has been made in loading GDExtensions which has some similarities.
+
+Godot uses [Emscripten](https://emscripten.org/) to compile C++ code to [Web Assembly](https://webassembly.org/) (WASM) that can be run as "native code" in web browsers. The Godot WASM acts as the main module and it's able to load other WASMs that are built as side modules, GDExtensions that are used in web exports are built as WASM side modules so Godot can load them dynamically.
```
```suggestion
Godot uses [Emscripten](https://emscripten.org/) to compile C++ code to [Web Assembly](https://webassembly.org/) (WASM) that can be run as "native code" in web browsers. The Godot WASM acts as the main module and it's able to load other WASMs that are built as side modules. GDExtensions that are used in web exports are built as WASM side modules so Godot can load them dynamically.
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+Desktop platforms have been supported since 4.0, but what some users may not know is that those platforms support all .NET runtimes. This is not new to 4.2 and, in fact, has been supported since 4.0. Godot's C# integration uses standard .NET so everything that works in a normal C# project should work in Godot too. To use a different runtime from the default (which is CoreCLR), users have to modif ...
+
+To enable NativeAOT in desktop platforms make sure to target .NET 7.0 or higher and set the `<PublishAOT>` property to true. Also, since Godot uses reflection, make sure to root the assemblies (this is done automatically for you when exporting to iOS because NativeAOT is currently the only supported runtime). Here's an example C# project:
```
```suggestion
To enable NativeAOT in desktop platforms, make sure to target .NET 7.0 or higher and set the `<PublishAOT>` property to true. Also, since Godot uses reflection, make sure to root the assemblies (this is done automatically for you when exporting to iOS because NativeAOT is currently the only supported runtime). Here's an example C# project:
```

JohnVeness commented on collections/_article/platform-state-in-csharp-for-godot-4-2.md:
```diff
@@ -0,0 +1,99 @@
+
+## What's next?
+
+As new releases of .NET become available, platform support gets better. NativeAOT support for mobile platforms is still experimental, and there's currently no support for the web platform. The .NET 9.0 release is set to include some improvements to NativeAOT and we may see initial support for the web platform, we'll see what makes it into the release in November 2024.
```
```suggestion
As new releases of .NET become available, platform support gets better. NativeAOT support for mobile platforms is still experimental, and there's currently no support for the web platform. The .NET 9.0 release is set to include some improvements to NativeAOT and we may see initial support for the web platform. We'll see what makes it into the release in November 2024.
```

raulsntos committed:
Apply suggestions from code review

Co-authored-by: John Veness <john.veness.github@pelago.org.uk>

raulsntos committed:
Attempt to clarify web section

raulsntos commented:
@MurderVeggie I've tried to improve the web section a bit. Let me know if you still think it's confusing. Here's a bit of a changelog:

- Tried to clarify that when referring to Godot supporting the Web platform I mean for **non-C#** projects. And that when referring to "building C# projects for the Web" I mean **non-Godot** C# projects.
- Removed the reference to progress being made in the GDExtension Web support since it's irrelevant for C# projects and may add confusion. Although I still try to explain why GDExtension can support the Web platform while Godot C# projects cannot because I think it's a question users will have.
- Added a small mention of future upstream improvements, even though that's mentioned in more detailed in the _What's next?_ section.

MurderVeggie commented:
Yes, I think it reads a lot clearer now, really well done!

coppolaemilio committed:
Add thumbnail and update excerpt/date

coppolaemilio committed:
Add css table style

shana commented:
> What about console platforms?

Putting on my W4 consoles lead 👒: C# support is a core part of our console ports roadmap.

coppolaemilio commented:
Thank you everyone!
